DEF-673012 | [Jenkins] Non-existing Test Sets in ALM are not taken into consideration in the job execution & US-671228 | [Jenkins Plug-in] Handle "Warning" status for tests running from ALM

Zsolt-Lukacs22 opened the pull request:
…to consideration in the job execution

Zsolt-Lukacs22 committed:
DEF-673012 | [Jenkins] Non-existing Test Sets in ALM are not taken into consideration in the job execution

Zsolt-Lukacs22 committed:
Not listing error in final summary; Correcting error cases and error message in case of non-existing tests

dorin7bogdan reviewed (changes requested):
...

dorin7bogdan commented:
> Please simplify the logic to detect the pipeline-mode then replace the text of the existing container.

Please ignore the previous message, it was intended for other PR.

Zsolt-Lukacs22 commented on HpToolsLauncher/Runners/AlmTestSetsRunner.cs:
```diff
@@ -1233,7 +1236,8 @@ public override TestSuiteRunResults Run()
             {
-                ConsoleWriter.WriteLine(Resources.AlmRunnerRunError + ex.Message);
+                ConsoleWriter.WriteErrLine(Resources.AlmRunnerRunError + ex.Message);
+                return null;
```
Will leave it like this, because there is no point to execute further after specific exceptions.

dorin7bogdan commented on HpToolsLauncher/Runners/AlmTestSetsRunner.cs:
```diff
@@ -1146,14 +1149,15 @@ public override TestSuiteRunResults Run()
-                Console.WriteLine("Empty target test set list");
+                ConsoleWriter.WriteErrLine("Empty target test set list");
+                Launcher.ExitCode = Launcher.ExitCodeEnum.Failed;
+                return null;
```
The line 1154 might be not needed. Please check and remove it if it's redundant.

Zsolt-Lukacs22 replied:
Removed.

Zsolt-Lukacs22 committed:
Removing unnecessary returns

Zsolt-Lukacs22 committed:
DEF-673012 | Handle the non existing Tests inside a Test Set

Zsolt-Lukacs22 committed:
US-671228 | [Jenkins Plug-in] Handle "Warning" status for tests running from ALM

Zsolt-Lukacs22 commented on HpToolsLauncher/Launcher.cs:
```diff
@@ -886,24 +886,30 @@ private void RunTests(IAssetRunner runner, string resultsFile, TestSuiteRunResul
+                    if ((numErrors <= 0) && (numFailures > 0) && (numSuccess > 0))
+                    {
+                        _exitCode = ExitCodeEnum.Unstable;
+                    }
```
If an error occurred through the job and its status is already set before to FAILED, it should remain FAILED and not overwritten. The same is valid for ABORTED state as well.

soleverxu replied:
It makes sense to check the exit code `Aborted` since once set it shall not be changed any more, however, for the `Failed` code, it shall not be checked here.

The `RunTests` function will call in two places. The first one is the normal run, which reflects the test run results. The second place is **rerun** and the `Failed` code may become `Unstable`. For example, if a test is failed first time and after rerun it is passed, then the final code shall be `Unstable` rather than `Failed`. Your changes may always set `Failed` in this scenario.

BTW, we also got an issue of the wrong exit code from customer on the standalone tools launcher repository, and a PR is committed to fix the exit code. You can check the changes for reference. https://github.com/MicroFocus/ADM-FT-ToolsLauncher/pull/69

Zsolt-Lukacs22 replied:
But how should we differentiate a situation when because a non-existing test set an error has occurred? I set the statement as is, to eliminate the possibility of overwriting such situation, but I completely agree with you on this point. If I modify the code, if an error has occurred of non-existing test set(s), it can be overwritten to `Unstable`.

Furthermore, I don't really understand the ambiguity between the two versions, between the one that I had previously here and the updated you mentioned in the comment. Should I leave it like this, with my statements, and without the `Failed` check?

Zsolt-Lukacs22 replied:
_Update_: only the `Aborted` state will not be overwritten.

soleverxu replied:
Looks good

Zsolt-Lukacs22 commented on HpToolsLauncher/Runners/AlmTestSetsRunner.cs:
```diff
@@ -1691,7 +1696,7 @@ private void UpdateCounters(TestRunResults test, TestSuiteRunResults testSuite)
         /// <param name="qcTestStatus"></param>
         /// <returns></returns>
-        private TestState GetTsStateFromQcState(string qcTestStatus)
+        private TestState GetTsStateFromQcState(TestExecStatus qcTestStatus)
```
Modified the function definition to better handle what is happening inside the function. We need the TestExecStatus instance inside the function body, to determine the correct status of the test.

Zsolt-Lukacs22 commented on HpToolsLauncher/Runners/AlmTestSetsRunner.cs:
```diff
@@ -1713,8 +1718,15 @@ private TestState GetTsStateFromQcState(string qcTestStatus)
+                case "FinishedPassed":
+					{
+                        if (qcTestStatus.Message.Contains("warning"))
+						{
```
If the ALM returned test message contains "warning" the test has generated a warning, but from ALM is wrongly returned as PASSED, we should overwrite it.

Zsolt-Lukacs22 commented on HpToolsLauncher/TestSuiteRunResults.cs:
```diff
@@ -53,6 +54,12 @@ public int NumTests
+        public int NumWarnings
+		{
+            get { return m_numWarnings; }
+            set { m_numWarnings = value; }
```
Number of generated warnings was not implemented on the TestSuiteRunResults side.

Zsolt-Lukacs22 commented on HpToolsLauncher/Runners/FileSystemTestsRunner.cs:
```diff
@@ -560,6 +561,9 @@ private void UpdateCounters(TestState testState)
                     _fail += 1;
                     break;
+                case TestState.Warning:
+                    _warnings += 1;
```
We should update the count of generated warnings.

Zsolt-Lukacs22 commented on HpToolsLauncher/Runners/AlmTestSetsRunner.cs:
```diff
@@ -1683,6 +1685,9 @@ private void UpdateCounters(TestRunResults test, TestSuiteRunResults testSuite)
                     ++testSuite.NumErrors;
                     break;
+                case TestState.Warning:
+                    ++testSuite.NumWarnings;
```
We should update the count of generated warnings.

soleverxu commented on HpToolsLauncher/Launcher.cs:
```diff
@@ -879,30 +879,37 @@ private void RunTests(IAssetRunner runner, string resultsFile, TestSuiteRunResul
                 //if there is an error
-                if (results.TestRuns.Any(tr => tr.TestState == TestState.Failed || tr.TestState == TestState.Error))
+                if (results.TestRuns.Any(tr => tr.TestState == TestState.Error))
                 {
```
Why removing the `TestState.Failed` in the condition? The `Failed` state shall also return Failed exit code, no?

soleverxu commented on HpToolsLauncher/Runners/AlmTestSetsRunner.cs:
```diff
@@ -1217,10 +1222,9 @@ public override TestSuiteRunResults Run()
-                //Launcher.ExitCode = Launcher.ExitCodeEnum.Failed;
-                Console.WriteLine(Resources.AlmTestSetsRunnerNoTestAfterApplyingFilters);
+                ConsoleWriter.WriteErrLine(Resources.AlmTestSetsRunnerNoTestAfterApplyingFilters);
+                Launcher.ExitCode = Launcher.ExitCodeEnum.Failed;
```
Do we really need to treat this scenario as a `Failed` state? 

Say, I'm a customer and I want to run all **unstable** ALM tests. So I do apply a status filter `unstable`. First time, I can get two unstable tests that will be run which is good. And then I run the job again, this time since there is no unstable test, no filtered tests are returned so eventually I received a **Failed** job result. It sounds not good for me. I may just want to run this job periodically to make sure there are no **unstable** tests left, otherwise, everything is good. However, your changes make it always return `Failed` exit code.

Zsolt-Lukacs22 replied:
We should only return `Failed` state in this case, if no filter is selected using the `IsFilterSelected` flag. Through the `FilterTests` method not only filters are applied, but tests are searched by name, if a specific test is given by the user, not a test set. Will modify.

Zsolt-Lukacs22 replied:
_Update_: will only return `Failed` state if a specific test is given by the user, otherwise won't.

soleverxu reviewed (changes requested):
See comments inline.

Zsolt-Lukacs22 committed:
Resolving exit code issues, only returning FAILED if no filters are applied

Zsolt-Lukacs22 committed:
Correcting error messages, moving FatalError check inside the ABORTED-check block, correcting logic of non-existing tests

In the case of non-existing tests full paths will be saved and they will always in the final summary.

Zsolt-Lukacs22 committed:
Making error messages homogenous

Zsolt-Lukacs22 committed:
Fixing bug of non-existing specific tests

If a specific test was given by the user, which were removed by the filter, we got non-existing case; this should not happen.
